ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -527,12 +527,66 @@ public func concat<T, E>(next: SignalProducer<T, E>)(producer: SignalProducer<T,
 	return SignalProducer(values: [producer, next]) |> concat
 }
 
+/// Merges a `producer` of SignalProducers down into a single producer, biased toward the producers
+/// added earlier. Returns a SignalProducer that will forward signals from the original producers
+/// as they arrive.
+public func merge<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> SignalProducer<T, E> {
+
+	let relay = SignalProducer<T, E> { relayObserver, relayDisposable in
+		
+		let inFlight = Atomic(1)
+		
+		let decrementInFlight: () -> () = {
+			let orig = inFlight.modify { $0 - 1 }
+			if orig == 1 {
+				sendCompleted(relayObserver)
+			}
+		}
+		
+		producer.startWithSignal { producerSignal, producerDisposable in
+			relayDisposable.addDisposable(producerDisposable)
+			
+			producerSignal.observe(next: { innerProducer in
+				innerProducer.startWithSignal { innerProducerSignal, innerProducerDisposable in
+					inFlight.modify { $0 + 1 }
+					
+					let innerProducerHandle = relayDisposable.addDisposable(innerProducerDisposable)
+					
+					let sink = SinkOf<Event<T, E>> { event in
+						if event.isTerminating {
+							innerProducerHandle.remove()
+						}
+						
+						switch event {
+						case .Completed:
+							decrementInFlight()
+							
+						default:
+							relayObserver.put(event)
+						}
+					}
+					
+					innerProducerSignal.observe(sink)
+				}
+				}, error: { error in
+					sendError(relayObserver, error)
+				}, completed: {
+					decrementInFlight()
+			})
+		}
+	}
+	
+	return relay
```

Comment:
Here too, RAC typically returns the `Signal(Producer)?` directly rather than assigning to a variable and returning the variable (assuming the variable is otherwise unused).